src/kOS.Safe/Encapsulation/PIDLoop.cs

```diff
@@ -187,84 +196,128 @@ public double Update(double sampleTime, double input, double setpoint, double ma
                 // because the error is within the epsilon:
-                error = 0;
+                Error = 0;
                 input = Setpoint;
                 Input = input;
-                Error = error;
             }
-            double pTerm = error * Kp;
-            double iTerm = 0;
-            double dTerm = 0;
+
+            PTerm = Error * Kp;
+
             if (LastSampleTime == double.MaxValue)
             {
-                lastIError = error * Ki;
+                lastIError = Error * Ki;
             }
             else if (LastSampleTime < sampleTime)
             {
                 double dt = sampleTime - LastSampleTime;
+
+                ChangeRate = (input - Input) / dt;
+                DTerm = -ChangeRate * Kd;
+
                 if (Ki != 0)
                 {
-                    if (ExtraUnwind)
+                    ExtraUnwindIfEnabled();
+
+                    ITerm += lastIError * dt;
+                    lastIError = AntiWindup(Error * Ki);
+                }
+                else
+                {
+                    ITerm = 0;
+                    lastIError = Error * Ki;
+                }
+            }
+            
+            LastSampleTime = sampleTime;
+            Input = input;
+            if (Ki != 0) ErrorSum = ITerm / Ki;
+            else ErrorSum = 0;
+
+            // Limit output according to MinOutput and MaxOutput
+            Output = Math.Min(MaxOutput, Math.Max(MinOutput, PTerm + ITerm + DTerm));
+            return Output;
+        }
+
+        private void ExtraUnwindIfEnabled()
+        {
+            if (ExtraUnwind)
+            {
+                if (Math.Sign(Error) != Math.Sign(ErrorSum))
+                {
+                    if (!unWinding)
                     {
-                        if (Math.Sign(error) != Math.Sign(ErrorSum))
-                        {
-                            if (!unWinding)
-                            {
-                                Ki *= 2;
-                                unWinding = true;
-                            }
-                        }
-                        else if (unWinding)
-                        {
-                            Ki /= 2;
-                            unWinding = false;
-                        }
+                        Ki *= 2;
+                        unWinding = true;
                     }
-                    iTerm = ITerm + lastIError * dt;
-                    lastIError = error * Ki;
                 }
-                ChangeRate = (input - Input) / dt;
-                if (Kd != 0)
+                else if (unWinding)
                 {
-                    dTerm = -ChangeRate * Kd;
+                    Ki /= 2;
+                    unWinding = false;
                 }
             }
-            else
+        }
+
+        private double AntiWindup(double iError)
+        {
+            double preSatOutput = PTerm + ITerm + DTerm;
+
+            switch (AntiWindupMode.ToUpper())
```

Comment:
If this value is going to be checked repeatedly during Update it's probably better to store it internally as an enum type and check the enum here instead of doing string manipulation on the fly again and again each Update().  (i.e. when a script uses ``set`` to change the value of the suffix, check it *then* to see if the string is one of the known values and if it is then set the enum, else throw an error.)  That way the string work only happens when the value is set, rather than 50 times a second.

Reply:
You're right, I'll fix this. Thanks!